extensions/amp-video-iframe/0.1/amp-video-iframe.js

```diff
@@ -0,0 +1,382 @@
+    this.canPlay_ = false;
+
+    /**
+     * @param {!Event} e
+     * @private
+     */
+    this.boundOnMessage_ = e => this.onMessage_(e);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  onLayoutMeasure() {
+    const {element} = this;
+    this.isAdLike_ = isAdLike(element);
+    this.isTrackingIframe_ = looksLikeTrackingIframe(element.getLayoutBox());
+  }
+
+  /** @override */
+  getLayoutPriority() {
+    if (this.isAdLike_) {
+      return LayoutPriority.ADS; // See AmpAd3PImpl.
+    }
+    if (this.isTrackingIframe_) {
+      return LayoutPriority.METADATA;
+    }
+    return super.getLayoutPriority();
+  }
+
+  /** @override */
+  buildCallback() {
+    installVideoManagerForDoc(this.getAmpDoc());
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.iframe_ = createFrameFor(this, this.getSrc_(), SANDBOX);
+    this.unlistenFrame_ = listen(this.win, 'message', this.boundOnMessage_);
+    return this.createReadyPromise_().then(() => this.onReady_());
+  }
+
+  /** @private */
+  onReady_() {
+    const {element} = this;
+    Services.videoManagerForDoc(element).register(this);
+    element.dispatchCustomEvent(VideoEvents.LOAD);
+  }
+
+  /** @override */
+  createPlaceholderCallback() {
+    const {element} = this;
+    const poster =
+        htmlFor(element)`<amp-img layout=fill placeholder></amp-img>`;
+
+    poster.setAttribute('src',
+        this.user().assertString(element.getAttribute('poster')));
+
+    return poster;
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    this.removeIframe_();
+    return true; // layout again.
+  }
+
+  /** @private */
+  removeIframe_() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    if (this.unlistenFrame_) {
+      this.unlistenFrame_();
+      this.unlistenFrame_ = null;
+      this.canPlay_ = false;
+    }
+  }
+
+  /** @private */
+  getSrc_() {
+    // TODO: assert https
+    return this.element.getAttribute('src');
+  }
+
+  /**
+   * @return {!Promise}
+   * @private
+   */
+  createReadyPromise_() {
+    const {promise, resolve, reject} = new Deferred();
+    this.readyPromise_ = promise;
+    this.readyResolver_ = resolve;
+    this.readyRejecter_ = reject;
+    return promise;
+  }
+
+  /**
+   * @param {!Event} event
+   * @private
+   */
+  originMatches_(event) {
+    return originMatches(event, this.iframe_, /.*/);
+  }
+
+  /**
+   * @param {!Event} event
+   * @private
+   */
+  onMessage_(event) {
+    if (!this.iframe_) {
+      return;
+    }
+
+    if (!this.originMatches_(event)) {
+      return;
+    }
+
+    const eventData = getData(event);
+    if (!isJsonOrObj(eventData)) {
+      return;
+    }
+
+    const data = objOrParseJson(eventData);
+
+    const messageId = data['id'];
+    const methodReceived = data['method'];
+
+    if (methodReceived) {
+      if (methodReceived == 'getIntersection') {
```

Comment:
I don't think we should support this. This would allow `auotplay` implementations that differs from ours.